File: libraries/python/trnrun/simulation.py
"""State for one simulation submitted through TRNRun Queue.

A simulation is mutated only while its manager is pumping queue output, on the
caller's own thread, so none of this state is synchronized. Runner outcomes and
queue completion metadata are retained without synthesizing events.
"""

from __future__ import annotations

from collections import Counter, deque
from pathlib import Path
from typing import Final

from trnrun.config import SimulationConfig
from trnrun.events import (
    ConfigEvent,
    LogEvent,
    ProgressEvent,
    QueueEvent,
    SettingEvent,
    StatusEvent,
    TrnRunEvent,
    is_terminal_status,
)

DEFAULT_MAX_LOG_EVENTS: Final[int] = 5000


class Simulation:
    """Event state for one queued simulation."""

    def __init__(
        self,
        deck_path: str | Path,
        config: SimulationConfig,
        sim_id: int,
        max_log_events: int = DEFAULT_MAX_LOG_EVENTS,
    ) -> None:
        """Initialize a pending simulation."""
        self.id: int = sim_id
        self.deck_path: Path = Path(deck_path)
        self.config: SimulationConfig = config

        self._accepted: bool = False
        self._finished: bool = False
        self._completion_event: QueueEvent | None = None
        self._status: StatusEvent | None = None
        self._progress: ProgressEvent | None = None
        self._config_event: ConfigEvent | None = None
        self._setting_event: SettingEvent | None = None
        self._logs: deque[LogEvent] = deque(maxlen=max_log_events)
        self._severity_counts: Counter[str] = Counter()

    def apply_event(self, event: TrnRunEvent) -> None:
        """Fold one runner event into the simulation state.

        Queue lifecycle events share the ``TrnRunEvent`` union but carry no
        runner state; the manager routes them to ``mark_accepted`` and
        ``mark_completed`` instead.
        """
        if self._finished:
            return

        match event:
            case StatusEvent():
                self._status = event
            case ConfigEvent():
                self._config_event = event
            case SettingEvent():
                self._setting_event = event
            case ProgressEvent():
                self._progress = event
            case LogEvent():
                self._logs.append(event)
                self._severity_counts[event.severity.lower()] += 1
            case QueueEvent():
                return

    def mark_accepted(self) -> None:
        """Record that a queue worker picked the simulation up."""
        self._accepted = True

    def mark_completed(self, event: QueueEvent) -> None:
        """Retain the queue's completion event and finish this run.

        Completion does not determine the outcome: success still requires
        the runner's own terminal ``STATUS/DONE`` event.
        """
        if self._finished:
            return
        self._completion_event = event
        self._finished = True

    @property
    def completion_event(self) -> QueueEvent | None:
        """Return QUEUE/COMPLETED metadata, or None if not received.

        The event's exit_code is None when the runner could not be launched.
        A missing event instead means the run has not been marked finished.
        """
        return self._completion_event

    @property
    def status(self) -> StatusEvent | None:
        """Return the latest status event."""
        return self._status

    @property
    def progress(self) -> ProgressEvent | None:
        """Return the latest progress event."""
        return self._progress

    @property
    def config_event(self) -> ConfigEvent | None:
        """Return the latest simulation configuration event."""
        return self._config_event

    @property
    def setting_event(self) -> SettingEvent | None:
        """Return the latest runner setting event."""
        return self._setting_event

    @property
    def logs(self) -> list[LogEvent]:
        """Return retained log events."""
        return list(self._logs)

    @property
    def is_running(self) -> bool:
        """Return whether the simulation is waiting or running."""
        return not self._finished

    @property
    def is_accepted(self) -> bool:
        """Return whether a queue worker picked the simulation up."""
        return self._accepted

    @property
    def is_finished(self) -> bool:
        """Return whether the queue reported completion for this run."""
        return self._finished

    @property
    def has_terminal_status(self) -> bool:
        """Return whether the runner reported a canonical terminal status."""
        return self._status is not None and is_terminal_status(self._status.status)

    @property
    def succeeded(self) -> bool:
        """Return whether a completed run has the terminal status `DONE`."""
        return self._finished and self._status is not None and self._status.status == "DONE"

    @property
    def log_count(self) -> int:
        """Return the total number of log events."""
        return sum(self._severity_counts.values())

    @property
    def notices(self) -> int:
        """Return the notice count."""
        return self._count_severity("notice")

    @property
    def warnings(self) -> int:
        """Return the warning count."""
        return self._count_severity("warn")

    @property
    def fatals(self) -> int:
        """Return the fatal count."""
        return self._count_severity("fatal")

    def _count_severity(self, prefix: str) -> int:
        """Count matching severities."""
        return sum(count for severity, count in self._severity_counts.items() if severity.startswith(prefix))

File: libraries/dotnet/src/TRNRun/Simulation.cs
namespace TRNRun;

/// <summary>Tracks state and events for one queued simulation.</summary>
/// <remarks>
/// The owning manager folds queue output into this object as it reads, so callers observe
/// updates live rather than through snapshots. State is updated synchronously on the
/// manager's thread and is not thread-safe. Queue completion, rather than terminal runner
/// status, finishes the simulation.
/// </remarks>
public sealed class Simulation
{
    private readonly List<LogEvent> _logs = [];

    /// <summary>Creates pending state for one submitted simulation.</summary>
    internal Simulation(string id, string deckPath, SimulationConfig config)
    {
        ArgumentException.ThrowIfNullOrWhiteSpace(id);
        ArgumentException.ThrowIfNullOrWhiteSpace(deckPath);
        ArgumentNullException.ThrowIfNull(config);

        Id = id;
        DeckPath = deckPath;
        Config = config;
        Logs = _logs.AsReadOnly();
    }

    /// <summary>Gets the case-sensitive queue request identifier.</summary>
    public string Id { get; }

    /// <summary>Gets the submitted deck path.</summary>
    public string DeckPath { get; }

    /// <summary>Gets the submitted simulation configuration.</summary>
    public SimulationConfig Config { get; }

    /// <summary>Gets whether a queue worker has accepted the simulation.</summary>
    public bool IsAccepted { get; private set; }

    /// <summary>Gets whether the queue has completed the simulation.</summary>
    public bool IsFinished => CompletionEvent is not null;

    /// <summary>Gets whether the simulation is waiting or running.</summary>
    public bool IsRunning => !IsFinished;

    /// <summary>Gets whether the latest runner status is terminal.</summary>
    public bool HasTerminalStatus =>
        Status?.Status is
            SimulationStatus.Done
            or SimulationStatus.Cancelled
            or SimulationStatus.Error
            or SimulationStatus.Timeout
            or SimulationStatus.Stalled;

    /// <summary>
    /// Gets whether the queue completed the simulation with status <see cref="SimulationStatus.Done"/>.
    /// </summary>
    /// <remarks>Exit codes and progress do not determine success.</remarks>
    public bool Succeeded => IsFinished && Status?.Status is SimulationStatus.Done;

    /// <summary>Gets the latest runner status event.</summary>
    public StatusEvent? Status { get; private set; }

    /// <summary>Gets the latest progress event.</summary>
    public ProgressEvent? Progress { get; private set; }

    /// <summary>Gets the latest native simulation configuration event.</summary>
    public ConfigEvent? ConfigEvent { get; private set; }

    /// <summary>Gets the latest effective runner settings event.</summary>
    public SettingEvent? SettingEvent { get; private set; }

    /// <summary>Gets the queue completion event.</summary>
    /// <remarks>
    /// A null event means the run is unfinished. A null exit code can indicate
    /// that the runner could not be launched.
    /// </remarks>
    public QueueEvent? CompletionEvent { get; private set; }

    /// <summary>Gets all retained log events, oldest first.</summary>
    /// <remarks>
    /// Logs are retained without limit. This is a live read-only view that grows as
    /// events arrive; copy it to take a stable snapshot.
    /// </remarks>
    public IReadOnlyList<LogEvent> Logs { get; }

    /// <summary>Gets the total number of received log events.</summary>
    public int LogCount => _logs.Count;

    /// <summary>Gets the number of notice events.</summary>
    public int Notices { get; private set; }

    /// <summary>Gets the number of warning events.</summary>
    public int Warnings { get; private set; }

    /// <summary>Gets the number of fatal events.</summary>
    public int Fatals { get; private set; }

    /// <summary>Records that a queue worker accepted the simulation.</summary>
    internal void MarkAccepted() => IsAccepted = true;

    /// <summary>Records queue completion for the simulation.</summary>
    internal void MarkCompleted(QueueEvent completion)
    {
        if (Accepts(completion))
        {
            CompletionEvent = completion;
        }
    }

    /// <summary>Applies a routed runner event to the simulation state.</summary>
    internal void ApplyRunnerEvent(TrnRunEvent runEvent)
    {
        if (!Accepts(runEvent))
        {
            return;
        }

        switch (runEvent)
        {
            case StatusEvent status:
                Status = status;
                break;

            case ProgressEvent progress:
                Progress = progress;
                break;

            case ConfigEvent config:
                ConfigEvent = config;
                break;

            case SettingEvent setting:
                SettingEvent = setting;
                break;

            case LogEvent log:
                RecordLog(log);
                break;
        }
    }

    /// <summary>Checks that an event belongs to this simulation and still applies to it.</summary>
    /// <remarks>A misrouted event, or any event after queue completion, is ignored.</remarks>
    private bool Accepts(TrnRunEvent runEvent) =>
        !IsFinished && string.Equals(runEvent.RunId, Id, StringComparison.Ordinal);

    /// <summary>Retains a log event and updates the severity counters.</summary>
    /// <remarks>
    /// Unrecognized severities are counted by <see cref="LogCount"/> alone.
    /// </remarks>
    private void RecordLog(LogEvent log)
    {
        _logs.Add(log);

        if (log.Severity.Equals("Notice", StringComparison.OrdinalIgnoreCase))
        {
            Notices++;
        }
        else if (log.Severity.Equals("Warning", StringComparison.OrdinalIgnoreCase))
        {
            Warnings++;
        }
        else if (log.Severity.Equals("Fatal", StringComparison.OrdinalIgnoreCase))
        {
            Fatals++;
        }
    }
}
